sunpy/map/mapbase.py

```diff
@@ -2501,6 +2502,68 @@ def _check_axes(self, axes, warn_different_wcs=False):
 
         return axes
 
+    def reproject_to(self, output_projection, algorithm='interp', **reproject_args):
+        """
+        Reproject the map
+
+        ..note::
+            This method requires the optional package `reproject` to be installed.
+
+        Parameters
+        ----------
+        output_projection : `dict` or `~astropy.wcs.WCS`
+            The destination FITS WCS header or WCS instance
+        algorithm : str
+            One of the supported `reproject` algorithms (see below)
+
+        Returns
+        -------
+        outmap : `~sunpy.map.GenericMap`
+            The reprojected map
+
+        Notes
+        -----
+        The supported `reproject` algorithms are:
+        * 'interp' for :func:`~reproject.reproject_interp`
+        * 'adaptive' for :func:`~reproject.reproject_adaptive`
+        * 'exact' for :func:`~reproject.reproject_exact`
+        See the respective documentation for these functions for additional keyword
+        arguments that are allowed.
+
+        If the output projection is specified by a `~astropy.wcs.WCS` instance, the
+        `shape_out` keyword argument must also be specified for the dimensions of the
+        output map.  The ordering of these dimensions is in NumPy order (i.e.,
+        secondary axis and then primary axis).
+        """
+        try:
+            import reproject
+        except ImportError:
+            raise ImportError("This method requires the optional package `reproject`.")
+
+        if not isinstance(output_projection, astropy.wcs.WCS):
+            output_projection = Header(output_projection)
+
+        # Select the desired reprojection algorithm
+        functions = {'interp': reproject.reproject_interp,
+                     'adaptive': reproject.reproject_adaptive,
+                     'exact': reproject.reproject_exact}
+        if algorithm not in functions:
+            raise ValueError(f"The specified algorithm must be one of: {list(functions.keys())}")
+        func = functions[algorithm]
+
+        # Block the `return_footpoint` keyword argument from being passed on
+        if 'return_footprint' in reproject_args:
+            del reproject_args['return_footprint']
+
+        # Reproject the array
+        output_array = func(self, output_projection, return_footprint=False, **reproject_args)
+
+        # Create and return a new GenericMap
+        if isinstance(output_projection, astropy.wcs.WCS):
+            output_projection = output_projection.to_header()
+        outmap = GenericMap(output_array, output_projection, plot_settings=self.plot_settings)
```

Comment:
Do we not want to feed the new map through the map factory so it picks up any source specific overridden properties that correctly interpret non-FITS compliant metadata?

Reply:
And should we also be copying across non-WCS metadata from the original map? (e.g., exposure time, detector name etc.)

Reply:
These are thorny considerations.  So far, I have simply approached it the same way as our existing gallery examples: don't use a source-specific subclass, and don't copy over non-WCS metadata.  If the non-WCS metadata is merged with the output header, it might work as intended with a source-specific subclass, but it could also blow up (e.g., if that subclass makes fixes to WCS information).